signer/src/network/libp2p/bootstrap.rs

```diff
@@ -0,0 +1,314 @@
+        tracing::debug!(?addresses, "handling pending outbound connection");
+        let addresses = addresses.to_vec();
+
+        // We're only interested in our outbound dialing activity. This _can_ be
+        // a listener if the swarm is attempting a hole-punch.
+        if !effective_role.is_dialer() {
+            return Ok(addresses);
+        }
+
+        for addr in &addresses {
+            if self.is_seed_address(addr) {
+                tracing::debug!(%addr, "attempting to dial seed address");
+            }
+        }
+
+        // Return the addresses as-is.
+        Ok(addresses)
+    }
+
+    fn on_swarm_event(&mut self, event: libp2p::swarm::FromSwarm) {
+        match event {
+            FromSwarm::ConnectionEstablished(e) => {
+                // We're only interested in our outbound activity.
+                if !e.endpoint.is_dialer() {
+                    return;
+                }
+
+                // If we've connected to ourselves, we don't need to do anything.
+                // This might happen if a node has its own address in the seed list.
+                if e.peer_id == self.local_peer_id() {
+                    return;
+                }
+
+                // Increment our total connection count. We're not interested
+                // in counting connections to ourselves.
+                self.connection_count = self.connection_count.saturating_add(1);
+
+                // Get the remote address, and remove the p2p protocol if it exists.
+                let mut addr = e.endpoint.get_remote_address().clone();
+                if matches!(addr.iter().last(), Some(Protocol::P2p(_))) {
+                    addr.pop();
+                }
+
+                // Check if the address is a seed address. If it's not, then
+                // we don't need to do anything so we return early.
+                if self.is_seed_address(&addr) {
+                    tracing::debug!(peer_id = %e.peer_id, %addr, "connected to seed");
+                } else {
+                    return;
+                }
+
+                // Update our connected seeds map.
+                match self.connected_seeds.entry(e.peer_id) {
+                    Entry::Occupied(mut entry) => {
+                        if entry.get_mut().insert(addr) {
+                            tracing::trace!("added connected seed: {:?}", e.peer_id);
+                        }
+                    }
+                    Entry::Vacant(entry) => {
+                        let mut set = HashSet::new();
+                        set.insert(addr);
+                        entry.insert(set);
+                    }
+                }
+            }
+            FromSwarm::ConnectionClosed(e) => {
+                // If this was a connection to ourselves then we don't need to do
+                // anything.
+                if e.peer_id == self.local_peer_id() {
+                    return;
+                }
```

Comment:
Shouldn't we also return if we are not the dialer, like above? Otherwise we would decrement `self.connection_count` even when it's some other remote disconnecting from us.

Reply:
I think I need to revisit this. The idea is that this behavior should kick in when the current connection count is zero, otherwise it doesn't need to do anything (and let the other behaviors i.e. Kademlia handle the rest). So I should probably be counting the number of connections based on a `HashSet<PeerId>` instead.